services/leader/service.go

```
@@ -0,0 +1,194 @@
+}
+
+// NewService creates a new leader service client based on an etcd client.
+func NewService(cli *clientv3.Client, opts Options) (services.LeaderService, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	return &multiClient{
+		clients:    make(map[string]*client),
+		opts:       opts,
+		etcdClient: cli,
+	}, nil
+}
+
+// Close closes all underlying election clients and returns all errors
+// encountered, if any.
+func (s *multiClient) Close() error {
+	s.Lock()
+	if s.closed {
+		s.Unlock()
+		return nil
+	}
+
+	s.closed = true
+	s.Unlock()
+
+	return s.closeClients()
+}
+
+func (s *multiClient) isClosed() bool {
+	s.RLock()
+	defer s.RUnlock()
+	return s.closed
+}
+
+func (s *multiClient) closeClients() error {
+	s.RLock()
+	errC := make(chan error, 1)
+	var wg sync.WaitGroup
+
+	for _, cl := range s.clients {
+		wg.Add(1)
+
+		go func(cl *client) {
+			if err := cl.close(); err != nil {
+				select {
+				case errC <- err:
+				default:
+				}
+			}
+			wg.Done()
+		}(cl)
+	}
+
+	s.RUnlock()
+
+	wg.Wait()
+	close(errC)
+
+	select {
+	case err := <-errC:
+		return err
+	default:
+		return nil
+	}
+}
+
+func (s *multiClient) getOrCreateClient(electionID string) (*client, error) {
+	s.RLock()
+	cl, ok := s.clients[electionID]
+	s.RUnlock()
+	if ok {
+		return cl, nil
+	}
+
+	clientNew, err := newClient(s.etcdClient, s.opts, electionID)
+	if err != nil {
+		return nil, err
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	cl, ok = s.clients[electionID]
+	if ok {
+		// another client was created between RLock and now, close new one
+		go clientNew.close()
+
+		return cl, nil
+	}
+
+	s.clients[electionID] = clientNew
+	return clientNew, nil
+}
+
+func (s *multiClient) Campaign(electionID string, opts services.CampaignOptions) (<-chan campaign.Status, error) {
+	if s.isClosed() {
+		return nil, errClientClosed
+	}
+
+	client, err := s.getOrCreateClient(electionID)
+	if err != nil {
+		return nil, err
+	}
+
+	if opts == nil {
```

Probably don't want to do that here and instead fail loudly.